modules/apps/callbacks/forwarding_test.go

```diff
@@ -0,0 +1,182 @@
+package ibccallbacks_test
+
+import (
+	"fmt"
+	"testing"
+
+	"github.com/stretchr/testify/suite"
+
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	"github.com/cosmos/ibc-go/modules/apps/callbacks/testing/simapp"
+	callbacktypes "github.com/cosmos/ibc-go/modules/apps/callbacks/types"
+	"github.com/cosmos/ibc-go/v8/modules/apps/transfer/types"
+	clienttypes "github.com/cosmos/ibc-go/v8/modules/core/02-client/types"
+	channeltypes "github.com/cosmos/ibc-go/v8/modules/core/04-channel/types"
+	ibctesting "github.com/cosmos/ibc-go/v8/testing"
+)
+
+func init() {
+	ibctesting.DefaultTestingAppInit = SetupTestingApp
+}
+
+type CallbacksForwardingTestSuite struct {
+	suite.Suite
+
+	coordinator *ibctesting.Coordinator
+
+	chainA *ibctesting.TestChain
+	chainB *ibctesting.TestChain
+	chainC *ibctesting.TestChain
+
+	pathAtoB *ibctesting.Path
+	pathBtoC *ibctesting.Path
+}
+
+// setupChains sets up a coordinator with 3 test chains.
+func (s *CallbacksForwardingTestSuite) setupChains() {
+	s.coordinator = ibctesting.NewCoordinator(s.T(), 3)
+	s.chainA = s.coordinator.GetChain(ibctesting.GetChainID(1))
+	s.chainB = s.coordinator.GetChain(ibctesting.GetChainID(2))
+	s.chainC = s.coordinator.GetChain(ibctesting.GetChainID(3))
+
+	s.pathAtoB = ibctesting.NewTransferPath(s.chainA, s.chainB)
+	s.pathBtoC = ibctesting.NewTransferPath(s.chainB, s.chainC)
+}
+
+func (s *CallbacksForwardingTestSuite) SetupTest() {
+	s.setupChains()
+
+	s.pathAtoB.Setup()
+	s.pathBtoC.Setup()
+}
+
+func TestIBCCallbacksForwardingTestsuite(t *testing.T) {
+	suite.Run(t, new(CallbacksForwardingTestSuite))
+}
+
+// TestForwardingWithMemoCallback tests that, when forwarding a packet with memo from A to B to C,
+// the callback is executed only on the final hop.
+// NOTE: this does not test the full forwarding behaviour (assert on amounts, packets, acks etc)
+// as this is covered in other forwarding tests.
+func (suite *CallbacksForwardingTestSuite) TestForwardingWithMemoCallback() {
+	testCases := []struct {
+		name                   string
+		testMemo               string
+		expCallbackMapOnChainB map[callbacktypes.CallbackType]int
```

Comment:
We discussed this with @chatton and our understanding is that when the memo specifies `src_callback` we expect the callback to be executed on B and not on C. basically when forwarding we ensure that all callbacks are executed on either the last hop (C) or the one immediately before that (B). 

If that's not the case, please let me know! I was surprised to see some callbacks executed on B but after some reasoning it started to make sense.

Reply:
yea very strange, but I guess not much we can do. Would be great to add into the docs somewhere? With multi packetdata, it should become more explicit when and where callbacks are executed (hopefully :sweat_smile:)

Reply:
I opened #6883 to ensure we document this properly